tensornetwork/tensornetwork.py

```diff
@@ -966,28 +966,36 @@ def contract_parallel(self, edge: Edge) -> Node:
       raise ValueError("Attempted to contract dangling edge: '{}'".format(edge))
     return self.contract_between(edge.node1, edge.node2)
 
-  def squeeze(self, edge: Edge) -> Node:
-    """Squeezes a dangling edge with unit size.
+  def squeeze(self, edge: Edge) -> Union[Node, Tuple[Node]]:
+    """Squeezes an edge with unit size.
+
+    If the edge is dangling it removes it from its node and updates the
+    corresponding tensor shape. If the edge is non-dangling it disconnects
+    and removes it.
 
     Args:
-      edge: Edge
+      edge: Edge to be squeezed (removed).
 
     Returns:
-      node: The node that had the squeezed edge.
+      node: The node (or nodes) that had the squeezed edge.
 
     Raises:
-      ValueError: If the given edge is not dangling.
       ValueError: If the given edge has non-unit size.
     """
-    if not edge.is_dangling():
-      raise ValueError(
-          "Attempted to squeeze non-dangling edge '{}'.".format(edge))
-
     node = edge.node1
     axis = edge.axis1
     if node.get_dimension(axis) > 1:
       raise ValueError(
           "Attempted to squeeze edge '{}' with non-unit size.".format(edge))
+
+    if not edge.is_dangling():
+      dangling1, dangling2 = self.disconnect(edge)
+      node1 = self.squeeze(dangling1)
+      node2 = self.squeeze(dangling2)
```

Comment:
Perhaps we should check the dimension of both axes before squeezing anything? Currently, if the axis of node2 (erroneously) has non-unit size, this may modify the network *before* throwing an exception.

Also, note that the dimension may be `None`.

Reply:
The size being `None` should not be an error - this occurs if the size of a dimension is dynamically determined during the execution of a graph and should be allowed. If the axis has dimension `None`, we should instead just let it go through and allow `tf.squeeze` to fail if the dynamic dimension was not 1.

Actually, I don't know if we should try to check the size at all. In `connect()` we do *not* check for compatible sizes, presumably because of the `None` case. Probably we should either check dimensions (where not `None`) in both places, or in neither.

Reply:
We should be checking sizes in both cases.

I think the best we can do is to do static checks outside of `tf.assert_equal` and just let `None`s slip by. Super not ideal, but having slow `Edge` creation is even worse.

Reply:
That would be my preference too. Check using `tensor.shape`. If `tensor.shape[i] is None` we let it pass.

We *could* add a `tf.assert_equal()` op only in case the shapes are undetermined, but I don't think it adds much... Also, one is supposed to explicitly make dependent graph ops depend on the assert op succeeding, which would be horrible to do within `TensorNetwork`.